// @lmportal/common/build/middlewares/urlfilter.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIfeatures = void 0;
var APIfeatures = /** @class */ (function () {
    function APIfeatures(query, queryString) {
        this.query = query;
        this.queryString = queryString;
    }
    APIfeatures.prototype.filter = function () {
        var querydObj = __assign({}, this.queryString);
        var excludedFields = [
            'page',
            'sort',
            'limit',
            'fields',
        ];
        excludedFields.forEach(function (el) { return delete querydObj[el]; });
        var queryStr = JSON.stringify(querydObj);
        queryStr = queryStr.replace(/\bgte|gt|lte|lt\b/g, function (match) { return "$" + match; });
        this.query = this.query.find(JSON.parse(queryStr));
        return this;
    };
    APIfeatures.prototype.sort = function () {
        if (this.queryString.sort) {
            var sortBy = this.queryString.sort
                .split(',')
                .join(' ');
            this.query = this.query.sort(sortBy);
        }
        else {
            this.query = this.query.sort('-createdAt');
        }
        return this;
    };
    APIfeatures.prototype.limitFields = function () {
        if (this.queryString.fields) {
            var fields = this.queryString.fields
                .split(',')
                .join(' ');
            this.query = this.query.select(fields);
        }
        else {
            this.query = this.query.select('-__v');
        }
        return this;
    };
    APIfeatures.prototype.paginate = function () {
        var limit = this.queryString.limit * 1 || 100;
        var page = this.queryString.page * 1 || 1;
        var skip = (page - 1) * limit;
        this.query = this.query.skip(skip).limit(limit);
        return this;
    };
    return APIfeatures;
}());
exports.APIfeatures = APIfeatures;
//module.exports = APIfeatures;
